internal/provider/kubernetes/gateway.go

```diff
@@ -67,12 +67,26 @@ func newGatewayController(mgr manager.Manager, cfg *config.Server, su status.Upd
 	if err := c.Watch(
 		&source.Kind{Type: &gwapiv1b1.Gateway{}},
 		&handler.EnqueueRequestForObject{},
-		predicate.NewPredicateFuncs(r.hasMatchingController),
+		predicate.NewPredicateFuncs(r.gatewayHasMatchingController),
 	); err != nil {
 		return err
 	}
 	r.log.Info("watching gateway objects")
 
+	// Only enqueue GatewayClass objects that match this Envoy Gateway's controller name.
+	if err := c.Watch(
+		&source.Kind{Type: &gwapiv1b1.GatewayClass{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.NewPredicateFuncs(r.classHasMatchingController),
+	); err != nil {
+		return err
+	}
+	r.log.Info("watching gatewayclass objects")
+
+	// Trigger Gateway reconciliation when a managed GatewayClass has changed.
+	if err := c.Watch(&source.Kind{Type: &gwapiv1b1.GatewayClass{}}, r.enqueueRequestForOwningGatewayClass()); err != nil {
+		return err
+	}
```

Comment:
I'm confused by this.

It's my understanding of controller-runtime that it's basically only ever correct to use `EnqueueRequestForObject` for one type within a single controller, but here you're adding using it for `GatewayClass` in addition to the existing use of it for `Gateway`.

And why are we setting up two separate watches for `GatewayClass` here?

And in that last one, isn't the relationship backwards? That it should be a `r.enqueueRequestsForOwnedGateways()` that takes a `GatewayClass` and lists `Gateways` that are associated with it?